File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0381_Insert_Delete_GetRandom_O1_Duplicates_allowed.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.*;

public class Question_0381_Insert_Delete_GetRandom_O1_Duplicates_allowed {
    private List<Integer> lst;
    private Map<Integer, Set<Integer>> idx;
    private Random rand = new Random();

    /** Initialize your data structure here. */
    public Question_0381_Insert_Delete_GetRandom_O1_Duplicates_allowed() {
        lst = new ArrayList<>();
        idx = new HashMap<>();
    }

    /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
    public boolean insert(int val) {
        if (!idx.containsKey(val)) {
            idx.put(val, new LinkedHashSet<>());
        }
        idx.get(val).add(lst.size());
        lst.add(val);
        return idx.get(val).size() == 1;
    }

    /** Removes a value from the collection. Returns true if the collection contained the specified element. */
    public boolean remove(int val) {
        if (!idx.containsKey(val) || idx.get(val).size() == 0) {
            return false;
        }
        int removeIdx = idx.get(val).iterator().next();
        idx.get(val).remove(removeIdx);
        int last = lst.get(lst.size() - 1);
        lst.set(removeIdx, last);
        idx.get(last).add(removeIdx);
        idx.get(last).remove(lst.size() - 1);

        lst.remove(lst.size() - 1);
        return true;
    }

    /** Get a random element from the collection. */
    public int getRandom() {
        return lst.get(rand.nextInt(lst.size()));
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0381_Insert_Delete_GetRandom_O1_Duplicates_allowed.cc
#include <stdlib.h>
#include <vector>
#include <set>
#include <unordered_map>
#include <iostream>

using namespace std;

class RandomizedCollection {
 public:
  /** Initialize your data structure here. */
  RandomizedCollection() {
  }
    
  /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
  bool insert(int val) {
    m[val].push_back(nums.size());
    nums.push_back(pair<int, int>(val, m[val].size() - 1));
    return m[val].size() == 1;
  }
    
  /** Removes a value from the collection. Returns true if the collection contained the specified element. */
  bool remove(int val) {
    if (m.find(val) == m.end()) {
      return false;
    }
    pair<int, int> last = nums.back();
    m[last.first][last.second] = m[val].back();
    nums[m[val].back()] = last;
    m[val].pop_back();
    if(m[val].empty()) {
      m.erase(val);
    }
    nums.pop_back();
    return true;
  }
    
  /** Get a random element from the collection. */
  int getRandom() {
    return nums[rand() % nums.size()].first;
  }

 private:
    // 值
    vector<pair<int, int>> nums;
    // 值 -> 值在nums中的索引
    unordered_map<int, vector<int>> m;
};
